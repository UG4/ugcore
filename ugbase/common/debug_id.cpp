/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#include "debug_id.h"
#include "common/log.h"
#include "common/error.h"
#include "common/assert.h"
#include <iostream>
#include <string.h>
#include "util/string_util.h"

namespace ug{

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of variables
 * of type DebugID, which are mostly global variables.
 * Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
DebugID::DebugID(const char *str)
{
	m_hash = crc32(str);
	GetDebugIDManager().register_debug_id(str);
}

DebugIDManager& DebugIDManager::instance()
{
	static DebugIDManager m;
	return m;
}


bool DebugIDManager::
set_debug_levels(int lev)
{
	for(std::map<uint32, int>::iterator it=m_dbgLevels.begin(); it != m_dbgLevels.end(); ++it)
		(*it).second = lev;
	return true;
}

bool DebugIDManager::
set_debug_level(const char *debugID, int level)
{
	int slen = strlen(debugID);
	if(slen<=0) return false;
	if(debugID[slen-1] == '*')
	{
		for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
		{
			const char *name = m_dbgLevelIdentifiers[i].c_str();
			if(WildcardMatch(name, debugID))
			{
				set_debug_level(crc32(name), level);
				//UG_LOGN(name);
			}
		}
	}
	else if(set_debug_level(crc32(debugID), level) == false)
	{
		UG_LOG("DebugID " << debugID << " not registered.\n");
		return false;
	}
	return true;
}

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of global variables
 * of type DebugID. Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
bool DebugIDManager::
register_debug_id(const char *debugID)
{
	if(debug_id_registered(debugID) == false)
	{
		m_dbgLevelIdentifiers.push_back(std::string(debugID));
		m_dbgLevels[crc32(debugID)] = -1;
		return true;
	}
	else
	{
		// not quite clear if cout is defined yet.
		// --> it should be, if we define it in this header!
		std::cout << "FATAL ERROR: DebugID "<<debugID<<" already registered." << std::endl;
		// note that this could be caused by double-registering libraries.
		UG_THROW("FATAL ERROR: DebugID "<<debugID<<" already registered.");
		return false;
	}
}

}
