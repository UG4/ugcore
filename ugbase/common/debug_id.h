/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef __H__UG__COMMON__DEBUG_ID_H
#define __H__UG__COMMON__DEBUG_ID_H
#include <vector>
#include <map>
#include <string>
#include "common/util/crc32.h"
#include <sstream>

namespace ug{

/// \addtogroup ugbase_common
/// \{

class DebugIDManager;



// DebugID
/**
 * A DebugID is an object to control the debug level of parts of the code.
 * It can be used in
 * - UG_SET_DEBUG_LEVEL
 * - UG_DEBUG_BEGIN/END (debugID, level)
 * - IF_DEBUG
 * - UG_DLOG(debugID, level, msg)
 *
 * ug4's standard DebugIDs are
 * APP, LIB_GRID, LIB_GRID_REFINER, LIB_DISC, LIB_DISC_ASSEMBLE, LIB_DISC_D3F,
 *		LIB_DISC_MULTIGRID, LIB_DISC_NEWTON, LIB_DISC_LINKER, LIB_DISC_TRANSFER,
 *		LIB_DISC_DISCRETE_FUNCTION, LIB_DISC_OUTPUT, LIB_DISC_OPERATOR_INVERSE,
 *		LIB_ALG_LINEAR_OPERATOR, LIB_ALG_LINEAR_SOLVER, LIB_ALG_VECTOR,
 *		LIB_ALG_MATRIX, LIB_ALG_AMG, LIB_PCL
 *
 * They are managed with \sa DebugIDManager.
 * You can define your own DebugID the following way:
 * In a cpp file:
 *
 \code{.cpp}
 DebugID DID_MYAPP("MyApp")
 DebugID DID_MYAPP_PARALLELIZATION("MyApp.Parallelization")
 \endcode{.cpp}
 *
 * Now to use it in other files you need to add there
 *
 \code{.cpp}
 extern DebugID DID_MYAPP, DID_MYAPP_PARALLELIZATION;
 ...
 UG_DLOG(DID_MYAPP, 3, "bla");
 \endcode{.cpp}
 *
 * Note that the DebugID OBJECT is called DID_MYAPP, but its string identifier
 * is "MyApp" (this is used in the script system). In LUA script you can do
 *
 * GetLogAssistant():set_debug_level("MyApp", 3)
 * or, with completion (UGIDE)
 * SetDebugLevel(debugID.MyApp, 3)
 *
 * \note The actual DebugID level is stored in DebugIDManager.
 *
 */
class DebugID
{
public:
/// registers the DebugID at DebugIDManager
	DebugID(const char *str);
	
	
	DebugID(uint32 hash) { m_hash = hash; }
/// returns the debug level via GetDebugIDManager.
	inline int get_debug_level() const;

/// sets the debug level via GetDebugIDManager.
	inline bool set_debug_level(int level);

	friend class DebugIDManager;
private:
	uint32 m_hash;
};

// the list of ug4's standard DebugIDs.
extern DebugID MAIN,
		APP,
		LIB_GRID,
		LIB_GRID_REFINER,
		LIB_DISC,
		LIB_DISC_ASSEMBLE,
		LIB_DISC_D3F,
		LIB_DISC_MULTIGRID,
		LIB_DISC_NEWTON,
		LIB_DISC_LINKER,
		LIB_DISC_TRANSFER,
		LIB_DISC_DISCRETE_FUNCTION,
		LIB_DISC_OUTPUT,
		LIB_DISC_OPERATOR_INVERSE,
		LIB_ALG_LINEAR_OPERATOR,
		LIB_ALG_LINEAR_SOLVER,
		LIB_ALG_VECTOR,
		LIB_ALG_MATRIX,
		LIB_ALG_AMG,
		LIB_PCL;


/**
 * The DebugIDManager.
 * get, set debug level, get a list of available DebugIDs etc.
 */
class DebugIDManager
{
private:
	/// returns the debug level of debugIDhash, -1 if not found
		inline int get_debug_level(uint32 debugIDhash) const
		{
			std::map<uint32, int>::const_iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return -1;
			else return (*it).second;
		}
		
		/// sets the debug level of debugIDhash if registered to 'level'
		inline bool set_debug_level(uint32 debugIDhash, int level)
		{
			std::map<uint32, int>::iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return false;
			else
			{
				(*it).second = level;
				return true;
			}
		}
public:
	///	returns a reference to the single instance of LogAssistant
		static DebugIDManager& instance();

	/// returns true if there is a DebugID with name debugIDname
		inline bool debug_id_registered(const char *debugIDname) const
		{
			return m_dbgLevels.find(crc32(debugIDname)) != m_dbgLevels.end();
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const DebugID &debugID) const
		{
			return get_debug_level(debugID.m_hash);
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const char *debugID) const
		{
			return get_debug_level(crc32(debugID));
		}
		
	/// 
		inline DebugID get_debug_id(const char *debugID) const
		{
			return DebugID(crc32(debugID));
		}
		
		inline DebugID get_or_create_debug_id(const char *debugID)
		{
			if(debug_id_registered(debugID)==false)
				register_debug_id(debugID);
			return DebugID(crc32(debugID));
		}
				

	/// sets the debug level of debugID if registered to 'level'
		inline bool set_debug_level(DebugID &debugID, int level)
		{
			return set_debug_level(debugID.m_hash, level);
		}
	/// sets the debug level of debugID if registered to 'level'
		bool set_debug_level(const char *debugID, int level);

	
	/// sets the debug level of all registered debugIDs to 'lev'
		bool set_debug_levels(int lev);

	/// returns a string describing all registered debugIDs.
		std::string get_registered_debug_IDs() const
		{
			std::stringstream str;
			str << "DebugIDs:\n";
			for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
				str << m_dbgLevelIdentifiers[i] << " : Level " << get_debug_level(m_dbgLevelIdentifiers[i].c_str()) << "\n";
			return str.str();
		}

		const std::vector<std::string> &get_registered_debug_IDs_arr() const
		{
			return m_dbgLevelIdentifiers;
		}



	bool register_debug_id(const char *debugID);

private:
	std::vector<std::string> m_dbgLevelIdentifiers;
	std::map<uint32, int> m_dbgLevels;
};

inline DebugIDManager& GetDebugIDManager()
{
	return DebugIDManager::instance();
}

inline int DebugID::get_debug_level() const
{
	return GetDebugIDManager().get_debug_level(*this);
}
inline bool DebugID::set_debug_level(int level)
{
	return GetDebugIDManager().set_debug_level(*this, level);
}

// TemporaryDebugLevel
/// a helper object to temporary set debug levels in a given scope
/**
 * A object of this class can be used to change a DebugID. The DebugID will
 * be re-set to the value before at destruction of the object.
 * This is especially useful if algorithms are calling other functions which
 * also have special DebugIDs levels. And you want to do a "debug debug".
 * example:
\code{.cpp}
  UG_SET_DEBUG_LEVEL(LIB_GRID, 1);

  // DebugID LIB_GRID is now at level 1
  {
  	  TemporaryDebugLevel tdl(LIB_GRID, 5);

  	  // DebugID LIB_GRID is now at level 5
  	  MyFunction();
  }

  // DebugID LIB_GRID is now again at 1.
\endcode
 * You can also specify differences:
 *
 \code{.cpp}
  TemporaryDebugLevel tdl(LIB_GRID, MY_DEBUG_ID, +1);
  // now LIB_GRID has the temporary debug level of MY_DEBUG_ID +1.
 */
#ifdef UG_ENABLE_DEBUG_LOGS
class TemporaryDebugLevel
{
	DebugID &did;

	int m_prevDebugLevel;
public:
	TemporaryDebugLevel(DebugID &debugID) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
	}

	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(temporaryDebugLevel);
	}

	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(debugSrc.get_debug_level()+diff);
	}

	/// resets the debugID to the previous value.
	~TemporaryDebugLevel()
	{
		did.set_debug_level(m_prevDebugLevel);
	}

	void set(int temporaryDebugLevel)
	{
		did.set_debug_level(temporaryDebugLevel);
	}
};
#else
class TemporaryDebugLevel
{

public:
	TemporaryDebugLevel(DebugID &debugID){}
	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel){}
	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0){}
	/// resets the debugID to the previous value.
	~TemporaryDebugLevel(){}
	void set(int temporaryDebugLevel){}
};
#endif

// end group ugbase_common
/// \}

}	// end namespace ug
#endif	/* __H__UG__COMMON__DEBUG_ID_H */

