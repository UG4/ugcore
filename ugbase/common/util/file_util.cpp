/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Torbjörn Klatt
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

/**
 * \file file_util.cpp
 * \date 2012-05-15
 * \brief Implementation of OS independent file utility functions
 */

#include "common/util/file_util.h"
#include "common/ug_config.h"
#include "common/error.h"
#include "common/assert.h"
#include "common/profiler/profiler.h"
#include "path_provider.h"
#include <vector>
#include <cstring>

using namespace std;

namespace ug
{
/// !!! Serial i/o version !!!
UG_API bool FileExists( const char *filename )
{
  PROFILE_FUNC(); // since i/o

  ifstream in( filename );
  if( in.good() ) {
    in.close();
    return true;
  }
  return false;
}

/// !!! Serial i/o version !!!
// UG_API size_t FileSize( const char *filename )
// {
//   PROFILE_FUNC(); // since i/o

//   if( !FileExists( filename ) ) {
//     UG_THROW( "The file " << filename << " could not be found." );
//   }

//   ifstream ifs;
//   ifs.open( filename, ios_base::in );

//   if( !ifs.good() ) {
//     UG_THROW( "The file " << filename << " could not be opened." );
//   }
//   ifs.seekg( 0, ios_base::end );
//   size_t length = ifs.tellg();
//   ifs.close();
//   return length;
// }

bool FileTypeIs( const char* filename, const char* extension )
{
	PROFILE_FUNC();
	std::string name( filename );
	size_t iExtPos = name.find_last_of(".");
	return ( iExtPos != std::string::npos && name.substr(iExtPos).compare(extension) == 0 );
}

/// !!! Serial i/o version !!!
UG_API bool FileCompare( const char *file1, const char *file2 )
{
  PROFILE_FUNC(); // since i/o

  // Make sure, both files do exist
  if( !FileExists( file1 ) || !FileExists( file2 ) ) {
    UG_THROW( "One or both files could not be found:" << file1 << ", " << file2 );
  }

  // If both files are one and the same, we don't need to compare them any more
  if( strcmp( file1, file2 ) == 0 ) {
    return true;
  }

  // If the files have different size, we know that they are different and can stop
  size_t size1 = FileSize( file1 ), size2 = FileSize( file2 );
  if( size1 != size2 ) {
    return false;
  }
  // If both files have zero size, we can stop as well
  if( size1 == 0 && size2 == 0 ) {
    return true;
  }

  // Open the files
  ifstream f1, f2;
  f1.open( file1, ios_base::in );
  f2.open( file2, ios_base::in );

  // This should be obsolete, as it is checked by FileExists, but we want to be
  // sure.
  if( !f1.good() || !f2.good() ) {
    UG_THROW( "One or both files could not be opened:" << file1 << ", " << file2 );
  }

  string line1 = "", line2 = "";
  bool diff = false;
  while( !f1.eof() || !f2.eof() ) {
    getline( f1, line1 );
    getline( f2, line2 );
    if( ( line1 != line2 ) ||
        ( f1.eof() && !f2.eof() ) ||
        ( !f1.eof() && f2.eof() ) ) {
      diff = true;
      break;
    }
  }

  f1.close();
  f2.close();

  return !diff;
}

/// !!! Serial i/o version !!!
/// in parallel, see ParallelReadFile.
bool ReadFile(const char* filename, vector<char> &file, bool bText)
{
  PROFILE_FUNC();
  
  size_t fileSize = FileSize(filename);

  FILE *f = fopen(filename, "rb");
  if(f==NULL) return false;


  if(bText)
    file.resize(fileSize + 1);
  else
    file.resize(fileSize);

  size_t readSize = fread(&file[0], 1, fileSize, f);
  UG_COND_THROW(readSize != fileSize,
  			  "Read mismatch in ReadFile. Wrong number of bytes read: "
          << readSize << ", expected: " << fileSize);

  fclose(f);
  if(bText) file[fileSize]=0x00;
  return true;
}

/// !!! Serial i/o version !!!
string MakeTmpFile(string filename, const string &extension, bool &bSuccess)
{
	PROFILE_FUNC();  // since i/o
	bSuccess = true;
	string t = filename+extension;
	const char *name = t.c_str();
	if(!FileExists(name)) return name;
	for(int i=0;i<999999; i++)
	{
		stringstream ss;
		ss << filename << i << extension;
		name = ss.str().c_str();
		if(!FileExists(name)) return name;
	}	
	bSuccess = false;
	return "";
}

////////////////////////////////////////////////////////////////////////////////
std::string FindFileInStandardPaths(const char* filename)
{
//	first check whether the file can be loaded (e.g. absolute path or relative to working-directory)
	std::string filenameOut = filename;
	if(FileExists(filenameOut.c_str()))
		return filenameOut;

//	Now check whether the file was specified relative to the current
//	scripting-directory or script or apps or root paths
	PathProvider::get_filename_relative_to_current_path(filename, filenameOut)
		|| PathProvider::get_filename_relative_to_path(SCRIPT_PATH, filename, filenameOut)
		|| PathProvider::get_filename_relative_to_path(APPS_PATH, filename, filenameOut)
		|| PathProvider::get_filename_relative_to_path(ROOT_PATH, filename, filenameOut);

	if(FileExists(filenameOut.c_str()))
		return filenameOut;

//	filename couldn't be located
	return "";
}


std::string FindDirInStandardPaths(const char* dirname)
{
//	first check whether the file can be loaded (e.g. absolute path or relative to working directory)
	std::string dirNameOut = dirname;
	if (DirectoryExists(dirNameOut.c_str()))
		return dirNameOut;

//	Now check whether the directory was specified relative to the current
//	scripting-directory or script or apps or root paths
	bool success = PathProvider::get_dirname_relative_to_current_path(dirname, dirNameOut)
		|| PathProvider::get_dirname_relative_to_path(SCRIPT_PATH, dirname, dirNameOut)
		|| PathProvider::get_dirname_relative_to_path(APPS_PATH, dirname, dirNameOut)
		|| PathProvider::get_dirname_relative_to_path(ROOT_PATH, dirname, dirNameOut);

	if (success)
		return dirNameOut;

//	filename couldn't be located
	return "";
}



} // namespace ug

// EOF
