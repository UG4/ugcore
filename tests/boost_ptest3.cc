
#define UG_PARALLEL

#include "lib_algebra/graph_interface/sparsematrix_boost.h"
#include "lib_algebra/cpu_algebra/sparsematrix_impl.h"

#include "lib_algebra/graph_interface/parallel_matrix.h"
#include "lib_algebra/graph_interface/parallel_matrix_boost.h"
#include "lib_algebra/graph_interface/bidirectional_boost.h"
#include "pcl/pcl_base.cpp"
#include "pcl/pcl_util.cpp"

// library?
#include "common/log.cpp"
#include "common/util/file_util.cpp"
#include "common/debug_id.cpp"
#include "common/assert.cpp"
#include "common/util/crc32.cpp"
#include "common/util/ostream_buffer_splitter.cpp"
#include "common/util/string_util.cpp"
#include "common/util/binary_buffer.cpp"
#include "common/util/os_dependent_impl/file_util_posix.cpp"
#include "common/util/os_dependent_impl/os_info_linux.cpp"

// #include "lib_algebra/parallelization/parallel_nodes.cpp" // ?

#include "pcl/pcl_process_communicator.cpp"
#include "pcl/pcl_comm_world.cpp"

#include "boost/graph/graph_utility.hpp"

#include "lib_algebra/parallelization/algebra_layouts.cpp" // operator<<
#include "lib_algebra/parallelization/parallel_index_layout.cpp"

// parallel_index_layout.h
// typedef pcl::SingleLevelLayout<pcl::OrderedInterface<size_t, std::vector> > IndexLayout;


typedef ug::SparseMatrix<double> T;
typedef ug::ParallelMatrix<T> PM;
typedef ug::BGLParallelMatrix<PM> BM;

template<class G>
class Map{
public:
	Map(G const& g):_g(g){}

	template<class T>
	std::string operator()(T& t) const{
//		using boost::detail::parallel::owner;
//		using boost::detail::parallel::local;
		return "(" + std::to_string(boost::owner(t)) + ":" + std::to_string(boost::local(t)) + ")";
	}
private:
	G const& _g;
};

template<class T, class G>
std::string get(Map<G> const& m, T t)
{
	return m(t);
}

int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	PM M;

	int rank;
	int size;
	MPI_Comm_rank(MPI_COMM_WORLD, &rank);
	MPI_Comm_size(MPI_COMM_WORLD, &size);
//	sleep(rank);

	static const int N=7 + 2*size;
	M.resize_and_clear(N, N);

	M(0, 0) = 2.;
	M(1, N-1) = -1.;
	for(int i=1; i<N; ++i){
		M(i, i) = 2.;
		M(i, i-1) = 1.;
		M(i-1, i) = 1.;
	}

	ug::AlgebraLayouts* L = new ug::AlgebraLayouts();
	pcl::ProcessCommunicator pc = L->proc_comm();
	IndexLayout& idx_master = L->master();
	IndexLayout& idx_slave = L->slave();

	// typedef pcl::OrderedInterface IndexLayout::Interface
	// pcl_communication_structs.h

	// build some arbitrary horizontal interfaces
	if(rank){ // "slave"
		IndexLayout::Interface& s = idx_slave.interface(0, 0);
		s.push_back(2*rank+1);
		s.push_back(2*rank+2);

	}else{ // master
		for(int p=1; p<size; ++p){
			IndexLayout::Interface& m = idx_master.interface(p, 0);
			m.push_back(2*p);
			m.push_back(2*p+1);
		}
	}

	{
		std::cout << "underlying...\n";
		auto e = boost::vertices(M);
		for(; e.first!=e.second; ++e.first){
			int v = *e.first;
			std::cout << " " << v;
		}
		std::cout << "\n";
	}

	M.set_layouts(make_sp(L));
	std::cout << *L <<"\n";

	BM b(&M);
	typedef typename boost::graph_traits<BM>::vertex_descriptor vertex_descriptor;
	typedef typename boost::graph_traits<BM>::out_edge_iterator out_edge_iterator;

	if(0)
	{
		auto e = boost::vertices(b);
		for(; e.first!=e.second; ++e.first){
			vertex_descriptor v = *e.first;
			std::cout << owner(v) << ":" << local(v) << " " << boost::out_degree(v, b) << "\n";
		}

		e = boost::vertices(b);
		for(; e.first!=e.second; ++e.first){
			vertex_descriptor v = *e.first;
			std::cout << owner(v) << ":" << local(v) << " " << boost::out_degree(v, b) << "\n";

//	auto wtmap = boost::get(boost::edge_weight, b);

#if 1 // need global vertex map or so
			std::cout << "=== out edges ===\n";
			std::pair<out_edge_iterator, out_edge_iterator> e = boost::out_edges(v, b);
			for(; e.first!=e.second; ++e.first){
				boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
				auto t = boost::target(edg, b);
				std::cout << local(boost::source(edg, b)) << " -- " << owner(t)
					<< ":" << local(t) << "\n";
			}
#endif
		}
	}

	std::cout << "pg\n";
	Map<BM> m(b);
	boost::print_graph(b, m);

	MPI_Finalize();
	
}

// BOOST_CONCEPT_ASSERT(( boost::DirectedGraphConcept<BM> ));
// BOOST_CONCEPT_ASSERT(( boost::DistributedGraphConcept<BM> )); // checker not available.
BOOST_CONCEPT_ASSERT(( boost::GraphConcept<BM> ));
BOOST_CONCEPT_ASSERT(( boost::AdjacencyGraphConcept<BM> ));
BOOST_CONCEPT_ASSERT(( boost::VertexListGraphConcept<BM> ));

typedef ug::BGLParallelMatrix<ug::ParallelMatrix<ug::SparseMatrix<ug::DenseMatrix<ug::FixedArray2<double, 3, 3> > > > > BMDF;
BOOST_CONCEPT_ASSERT(( boost::VertexListGraphConcept<BMDF> ));
