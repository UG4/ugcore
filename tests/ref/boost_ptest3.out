underlying...
 0 1 2 3 4 5 6 7 8
AlgebraLayouts:
 master: IndexLayout: 
 slave: IndexLayout: 
 vertical master: IndexLayout: 
 vertical slave: IndexLayout: 
 process communicator: ProcessCommunicator (WORLD) size = 1. involved procs: [ 0 ]

pg
(0:0) --> (0:0) (0:1) 
(0:1) --> (0:0) (0:1) (0:2) (0:8) 
(0:2) --> (0:1) (0:2) (0:3) 
(0:3) --> (0:2) (0:3) (0:4) 
(0:4) --> (0:3) (0:4) (0:5) 
(0:5) --> (0:4) (0:5) (0:6) 
(0:6) --> (0:5) (0:6) (0:7) 
(0:7) --> (0:6) (0:7) (0:8) 
(0:8) --> (0:7) (0:8) 
